File: Week 2/id_545/LeetCode_144_545.js
// 144. 二叉树的前序遍历

/*
// 迭代法
var preorderTraversal = function(root) {
    let arr = []
    let stack = []
    let curr = root
    while (curr || stack.length != 0) {
        if (curr) {
            arr.push(curr.val)
            stack.push(curr)
            curr = curr.left
        } else {
            curr = stack.pop()
            curr = curr.right
        }
    }
    return arr
};
*/

// 递归
class Preorder {
    constructor() {
        this.arr = []
    }
    traversal (root) {
        if (root) {
            this.arr.push(root.val)
            this.traversal(root.left)
            this.traversal(root.right)
        }
        return this.arr
    }
}
var preorderTraversal = function(root) {
    let preor = new Preorder()
    return preor.traversal(root)
}

File: Week 2/id_545/LeetCode_144_545.py
## 144. 二叉树的前序遍历

# 递归
# class Solution:
#     def __init__(self):
#         self.arr = []
        
#     def preorderTraversal(self, root: TreeNode) -> List[int]:
#         if root:
#             self.arr.append(root.val)
#             self.preorderTraversal(root.left)
#             self.preorderTraversal(root.right)
        
#         return self.arr


# 迭代法
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        arr = []
        stack = []
        curr = root
        while curr or stack:
            if curr:
                arr.append(curr.val)
                stack.append(curr)
                curr = curr.left
            else:
                curr = stack.pop()
                curr = curr.right
                
        return arr